### PR TITLE
ONNX Export Scatter

### PR DESCRIPTION
**Pull request by lara-hdr**

### CODE REVIEW

**Review by houseroad — COMMENTED** [review]
Could you add test cases?

**Comment by lara-hdr** [comment]
Caffe2 does not have an operator Scatter, but has ScatterAssign defined in utility_ops (that enforces in place operations). Will I have to add a new operator Scatter?

**Comment by houseroad** [comment]
Yeah, they are not quite aligned. I would suggest to add a new Scatter operator then.

**Comment by houseroad** [comment]
cc: @BIT-silence for the scatter op part.

File: caffe2/operators/utility_ops.h [hunk 1]
```diff
@@ -738,6 +738,122 @@ class ScatterAssignOp : public Operator<Context> {
...
+    const auto indicesType = TypeMetaToDataType(indices.dtype());
+    auto* output = Output(0);
+
+    auto runner = GetRunner(dataType, slicesType, indicesType);
+    (this->*runner)();
+    return true;
+  }
+
+ private:
+  typedef void (ScatterOp::*RunnerType)();
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]: I'd suggest to use using instead of typedef

File: caffe2/operators/utility_ops.h [hunk 2]
```diff
@@ -738,6 +738,122 @@ class ScatterAssignOp : public Operator<Context> {
...
+  typedef std::
+      map<std::pair<TensorProto_DataType, TensorProto_DataType>, RunnerType>
+          RunnerMap;
+
+  RunnerMap runners_;
+
+  RunnerType GetRunner(
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]: Please declare member functions before member variables.
https://google.github.io/styleguide/cppguide.html#Declaration_Order

File: caffe2/operators/utility_ops.h [hunk 3]
```diff
@@ -738,6 +738,122 @@ class ScatterAssignOp : public Operator<Context> {
...
+      const TensorProto_DataType dataType,
+      const TensorProto_DataType slicesType,
+      const TensorProto_DataType indicesType) {
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]:
```suggestion
      const TensorProto_DataType indicesType) const {
```

File: caffe2/operators/utility_ops.h [hunk 4]
```diff
@@ -738,6 +738,122 @@ class ScatterAssignOp : public Operator<Context> {
...
+    CAFFE_ENFORCE_EQ(dataType, slicesType, "Data and slice types must match");
+    auto it = runners_.find({indicesType, dataType});
+    CAFFE_ENFORCE(
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]:
```suggestion
    CAFFE_ENFORCE_NE(
```

File: caffe2/operators/utility_ops.h [hunk 5]
```diff
@@ -738,6 +738,122 @@ class ScatterAssignOp : public Operator<Context> {
...
+        it != runners_.end(),
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]:
```suggestion
        it, runners_.end(),
```

File: caffe2/operators/utility_ops.h [hunk 6]
```diff
@@ -738,6 +738,122 @@ class ScatterAssignOp : public Operator<Context> {
...
+        "Could not find the runner corresponding to indicesType, dataType = ",
+        indicesType,
+        " ",
+        dataType);
+    return it->second;
+  }
+
+  template <typename Index, typename T>
+  void DoRun() {
+    auto& input = Input(DATA);
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]:
```suggestion
    const auto& input = Input(DATA);
```

File: caffe2/operators/utility_ops.h [hunk 7]
```diff
@@ -738,6 +738,122 @@ class ScatterAssignOp : public Operator<Context> {
...
+    auto& indices = Input(INDICES);
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]:
```suggestion
    const auto& indices = Input(INDICES);
```

File: caffe2/operators/utility_ops.h [hunk 8]
```diff
@@ -738,6 +738,122 @@ class ScatterAssignOp : public Operator<Context> {
...
+    auto& slices = Input(SLICES);
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]:
```suggestion
    const auto& slices = Input(SLICES);
```

File: caffe2/operators/utility_ops.h [hunk 9]
```diff
@@ -738,6 +738,122 @@ class ScatterAssignOp : public Operator<Context> {
   INPUT_TAGS(DATA, INDICES, SLICES);
 };
 
+template <class Context>
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]: I'd suggest to make this op in a separate cc and h file. It is not a very good design to put so many ops in utility_ops.

**Comment by houseroad** [comment]
@lara-hdr could you address the comments from @BIT-silence ?

**Comment by lara-hdr** [comment]
@houseroad conflicts are solved in this PR :)

**Comment by lara-hdr** [comment]
@houseroad, comments on this PR?

**Comment by lara-hdr** [comment]
@houseroad updates on this?

File: caffe2/operators/utility_ops.h [hunk 10]
```diff
@@ -738,6 +738,105 @@ class ScatterAssignOp : public Operator<Context> {
...
+  template <typename Index>
+  bool DoRunWithType() {
+    const Tensor& data = Input(DATA);
+    const Tensor& indices = Input(INDICES);
+    const Tensor& updates = Input(UPDATES);
+    const TypeMeta dataType = data.dtype();
+    size_t item_bytesize = dataType.itemsize();
+
+    // ONNX allows negative axis to index from the back, valid range: [-r, r].
+    if (axis_ < 0) {
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]:
```suggestion
    axis_ = data.canonical_axis_index(axis_);
```

File: caffe2/operators/utility_ops.h [hunk 11]
```diff
@@ -738,6 +738,105 @@ class ScatterAssignOp : public Operator<Context> {
...
+  }
+
+  virtual ~ScatterOp() noexcept {}
+
+  bool RunOnDevice() override {
+    return DispatchHelper<TensorTypes<int32_t, int64_t>>::call(
+        this, this->template Input<Tensor>(INDICES, CPU));
+  }
+
+  template <typename Index>
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]: Please use TIndex or IndexType instead of Index.

File: caffe2/operators/utility_ops.h [hunk 12]
```diff
@@ -738,6 +738,105 @@ class ScatterAssignOp : public Operator<Context> {
...
+      axis_ = data.dim() + axis_;
+    }
+    
+    CAFFE_ENFORCE_GE(data.dim(), axis_ + 1, "DATA should be at least [axis+1]-D");
+    CAFFE_ENFORCE_GE(axis_, 0, "Axis should be non-negative");
+    CAFFE_ENFORCE_LT(axis_, data.dim(), "Axis out of range");
+
+    Tensor* output = Output(0, data.sizes().vec(), at::dtype(dataType));
+    output->CopyFrom(data);
+    auto out = static_cast<char*>(output->raw_mutable_data(dataType));
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]:
```suggestion
    char* out = static_cast<char*>(output->raw_mutable_data(dataType));
```
For pointer type, please use auto* or real type* to indicate this var is a pointer.

File: caffe2/operators/utility_ops.h [hunk 13]
```diff
@@ -738,6 +738,105 @@ class ScatterAssignOp : public Operator<Context> {
...
+    // Succeed if size of output is zero, which can happen for empty batch which
+    // would have data dimension size of 0.
+    // This *must* be done AFTER output->raw_mutable_data() above as that has
+    // important allocation side effect that we must see.
+    if (output->numel() == 0) {
+      return true;
+    }
+
+    const Index* idxs = indices.template data<Index>();
+    auto src_base = static_cast<const char*>(updates.raw_data());
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]:
```suggestion
    const char* src_base = static_cast<const char*>(updates.raw_data());
```

File: caffe2/operators/utility_ops.h [hunk 14]
```diff
@@ -738,6 +738,105 @@ class ScatterAssignOp : public Operator<Context> {
...
+  explicit ScatterOp(Args&&... args)
+      : Operator<Context>(std::forward<Args>(args)...),
+        OP_SINGLE_ARG(int, "axis", axis_, 1) {
+  }
+
+  virtual ~ScatterOp() noexcept {}
+
+  bool RunOnDevice() override {
+    return DispatchHelper<TensorTypes<int32_t, int64_t>>::call(
+        this, this->template Input<Tensor>(INDICES, CPU));
```
Comment (BIT-silence) [inline caffe2/operators/utility_ops.h]: Should we fix CPU here?

File: caffe2/operators/utility_ops.h [hunk 15]
```diff
@@ -738,6 +738,105 @@ class ScatterAssignOp : public Operator<Context> {
...
+ public:
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+
+  template <class... Args>
+  explicit ScatterOp(Args&&... args)
+      : Operator<Context>(std::forward<Args>(args)...),
+        OP_SINGLE_ARG(int, "axis", axis_, 1) {
+  }
+
+  virtual ~ScatterOp() noexcept {}
```
Comment (houseroad) [inline caffe2/operators/utility_ops.h]: In file included from caffe2/caffe2/operators/utility_ops.cc:1:
caffe2/caffe2/operators/utility_ops.h:752:11: warning: '~ScatterOp' overrides a destructor but is not marked 'override'

```
virtual ~ScatterOp() noexcept {}
        ^
caffe2/c10/util/Type.h:19:56: note: in instantiation of template class 'caffe2::ScatterOp<caffe2::CPUContext>' requested here
  static const auto& name = *(new std::string(demangle(typeid(T).name())));
                                                       ^
caffe2/caffe2/operators/utility_ops.cc:54:1: note: in instantiation of function template specialization 'c10::demangle_type<caffe2::ScatterOp<caffe2::CPUContext> >' requested here
REGISTER_CPU_OPERATOR(Scatter, ScatterOp<CPUContext>);
^
caffe2/caffe2/core/operator.h:1257:3: note: expanded from macro 'REGISTER_CPU_OPERATOR'
  C10_REGISTER_CLASS(CPUOperatorRegistry, name, __VA_ARGS__)
  ^
caffe2/c10/util/Registry.h:291:3: note: expanded from macro 'C10_REGISTER_CLASS'
  C10_REGISTER_TYPED_CLASS(RegistryName, #key, __VA_ARGS__)
  ^
caffe2/c10/util/Registry.h:251:14: note: expanded from macro 'C10_REGISTER_TYPED_CLASS'
      ::c10::demangle_type<__VA_ARGS__>());
             ^
caffe2/caffe2/core/operator.h:751:3: note: overridden virtual function is here
  ~Operator() noexcept override {}
  ^

```

Comment (houseroad) [reply caffe2/operators/utility_ops.h]: please address this comment as well, thanks.

File: caffe2/operators/utility_ops.h (hunk 14 above)
Comment (houseroad) [reply caffe2/operators/utility_ops.h]: I would suggest to provide better information here, if the tensor is on some devices other than CPU. To explicitly tell them, Scatter only supports CPU at the moment. :-)

Comment (BIT-silence) [reply caffe2/operators/utility_ops.h]: If it is a CPU only op, why not remove template Context?

Comment (lara-hdr) [reply caffe2/operators/utility_ops.h]: added a check on the device type, and changed Operator<Context> to Operator<CPUContext>.

**Review by houseroad — APPROVED** [review]
LGTM, thanks.

File: torch/onnx/symbolic_opset9.py [hunk 16]
```diff
@@ -1542,6 +1542,24 @@ def argmin(g, input, dim, keepdim):
         return g.op('ArgMin', input, axis_i=dim, keepdims_i=keepdim)
 
 
+@parse_args('v', 'i', 'v', 'v')
+def scatter(g, self, dim, index, src):
+    return g.op("Scatter", self, index, src, axis_i=dim)
+
+
+@parse_args('v', 'i', 'v', 'v')
+def scatter_add(g, self, dim, index, src):
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: Add a test?

**Comment by houseroad** [comment]
Feel free to add a test for `scatter_add` in a different PR, since this one is already closed. :-)

**Pull request closed without merging**